Firestore/core/src/model/overlayed_document.h

```diff
@@ -40,6 +40,10 @@ class OverlayedDocument {
     return document_;
   }
 
+  model::Document consume_document() {
```

Comment:
Was there a reason not to just overload `document()` on rvalue reference?

Reply:
The truth is I had problems to make it work when I tried. I tried again today and it worked..but it got me curious so I checked style guide: go/cstyle#Rvalue_references.

My read is that we should favor pass/return values instead of rvalues unless we can prove this helps with performance? I also think `consume_document()` states the intent better.

Reply:
I tend to disagree. I'll just give my rationale and you can make the final decision.

go/cstyle#Rvalue_references doesn't address overloading based on whether or not the method is invoked on an rvalue reference. It's talking more about method _arguments_.

Overloading on an rvalue reference makes it _automatic_ to move the object instead of requiring the caller at the call site to be sure to invoke the correct method. The compiler will pick the `&&` overload if invoked on a temporary object and the `const&` method if invoked on a non-temporary object.

Example 1: This will call the `const&` overload of `document()`:

```
void MyFunc(const OverlayedDocument& overlayed_document) {
  printf("The overlayed document is: %s", overlayed_document.document());
}
```

Example 2: This will call the `&&` overload of `document()`:
```
void MyFunc(OverlayedDocument&& overlayed_document) {
  overlayed_documents_.push_back(std::move(overlayed_document).document());
}
```

Example 2 is effectively the same as calling

```
overlayed_documents_.push_back(overlayed_document.consume_document());
```

because `consume_document()` will return a temporary object; however, the std::move notation is standard C++11 notation instead of writing a specially-named method. This same notation can be used to move objects out of absl::optional objects:

```
void SnapshotCreator::UpdateSnapshot(absl::optional<OverlayedDocument> doc) {
  if (doc.has_value()) {
    docs_.push_back(std::move(doc).value());
  }
}
```

In this example, `absl::optional` overloads the `value()` method on `&&` to move the object out of the optional if invoked on a temporary object. There is no special "consume_value()" method.

Reply:
I just pushed up a commit to do this.